ksqldb-execution/src/main/java/io/confluent/ksql/schema/query/QuerySchemas.java

```diff
@@ -118,47 +121,55 @@ public synchronized void trackSerdeOp(
   }
 
+  Map<String, Map<Boolean, Set<String>>> getTopicsToLoggers() {
+    return Collections.unmodifiableMap(topicsToLoggers);
+  }
+
   /**
-   * Builds complete SchemaInfo for a topic:
+   * Returns all different schemas and serde formats that the {@code topicName} is using. The
+   * {@link MultiSchemaInfo} contains a list of key/value schemas and serde formats that were
+   * tracked for the specified topic.
+   * </p>
+   * Key and/or value schemas may be empty if no schemas were detected.
    */
-  public SchemaInfo getTopicInfo(final String topicName) {
+  public MultiSchemaInfo getTopicInfo(final String topicName) {
+    final Set<SchemaInfo> keySchemasInfo = getTopicSchemas(topicName, IS_KEY_SCHEMA);
+    final Set<SchemaInfo> valueSchemasInfo = getTopicSchemas(topicName, IS_VALUE_SCHEMA);
+
+    return new MultiSchemaInfo(keySchemasInfo, valueSchemasInfo);
+  }
+
+  /**
+   * Returns a list of schemas and serde formats detected on the specified topicName.
+   * The isKeySchema is a boolean value that specifies whether to look at the key schema (True)
+   * or at the value schema (False).
+   * </p>
+   * Each schema information is tracked by the loggers the topicName has. The logger is
+   * a changelog topic. Some topics may have multiple schemas and serde formats if the topicName
+   * is used with stream-stream joins and foreign key joins.
+   */
+  private Set<SchemaInfo> getTopicSchemas(final String topicName,
+                                          final boolean isKeySchema) {
+    // Look at the different changelog topics that tne topicName uses. There might be multiple
+    // changelog topics if the topicName was used by joins, which contains internal state stores
+    // with a changelog per store.
     final Map<Boolean, Set<String>> kvLoggerNames = topicsToLoggers.get(topicName);
     if (kvLoggerNames == null) {
       throw new IllegalArgumentException("Unknown topic: " + topicName);
     }
 
-    final Set<String> keyLoggerNames = kvLoggerNames.getOrDefault(true, ImmutableSet.of());
-    if (keyLoggerNames.size() != 1) {
-      throw new IllegalStateException("Multiple key logger names registered for topic."
-          + System.lineSeparator()
-          + "topic: " + topicName
-          + "loggers: " + keyLoggerNames
-      );
-    }
-
-    final Set<String> valueTopicNames = kvLoggerNames.getOrDefault(false, ImmutableSet.of());
-    if (valueTopicNames.size() != 1) {
-      throw new IllegalStateException("Multiple value logger names registered for topic."
-          + System.lineSeparator()
-          + "topic: " + topicName
-          + "loggers: " + valueTopicNames
-      );
-    }
+    final Map<SchemaInfo, Set<String>> schemaToLoggers = new HashMap<>();
 
-    final SchemaInfo keyInfo = loggerToSchemas.get(Iterables.getOnlyElement(keyLoggerNames));
-    final SchemaInfo valueInfo = loggerToSchemas.get(Iterables.getOnlyElement(valueTopicNames));
-    if (keyInfo == null || valueInfo == null) {
```

Comment:
Was this check removed intentionally? It still seems good to throw if either the key or value info is missing.

Reply:
It's not necessary anymore. Now  the key and value info are a `Set` which maybe empty. The `getTopicSchemas` will never return null.
```
final Set<SchemaInfo> keySchemasInfo = getTopicSchemas(topicName, IS_KEY_SCHEMA);
final Set<SchemaInfo> valueSchemasInfo = getTopicSchemas(topicName, IS_VALUE_SCHEMA);
```

Reply:
Right but previously the test framework ensured that there is always key and value info for every topic, i.e., that you never have a topic with key info but no value info or vice versa. The version of the code does not enforce this anymore. Do you not think this is worth enforcing still?

Reply:
It is now enforced in the `TopicInfoCache`. It is now up to the caller to check now, which up to now only `TopicInfoCache` calls it. The idea now is that `getTopicInfo` just returns schema and formats found without throwing exceptions (except when the topic is unknown).

Reply:
Oh I see, it's being enforced indirectly through `TopicInfoCache` since `TopicInfoCache` requires that a key format and at least one value format exist, and those formats come from the schemas. I didn't realize the second part until now. Thanks for clarifying!